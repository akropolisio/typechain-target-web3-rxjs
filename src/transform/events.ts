import { Contract, EventDeclaration, getSignatureForFn } from 'typechain';

import { codegenInputTypes } from './types';

export function codegenEvents(events: Contract['events']): string {
  return Object.values(events)
    .map(evs => {
      if (evs.length === 1) {
        return codegenForSingleEvent(evs[0]);
      }
      return codegenForOverloadedEvents(evs);
    })
    .join('\n');
}

function codegenForOverloadedEvents(events: EventDeclaration[]): string {
  return [
    codegenForSingleEvent(events[0]),
    ...events.map(e => codegenForSingleEvent(e, `'${getSignatureForFn(e as any)}'`)),
  ].join('\n');
}

function codegenForSingleEvent(event: EventDeclaration, overloadedName?: string): string {
  return `${overloadedName ?? event.name}: EventMethod<
    ${codegenInputTypes(event.inputs)},
    ${codegenInputTypes(event.inputs.filter(e => e.isIndexed))}
  >`;
}

export function codegenPastEvents(events: Contract['events']): string {
  return Object.values(events)
    .flat()
    .map(x => x.name)
    .join(';\n'); // TODO unmock
}

function codegenFor