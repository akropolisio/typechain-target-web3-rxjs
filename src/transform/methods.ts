import { Contract, FunctionDeclaration, getSignatureForFn } from 'typechain';

import { codegenInputTypes, codegenOutputTypes } from './types';

export function codegenMethods(functions: Contract['functions']): string {
  return Object.values(functions)
    .map(fns => {
      if (fns.length === 1) {
        return codegenForSingleFunction(fns[0]);
      }
      return codegenForOverloadedFunctions(fns);
    })
    .join('\n');
}

function codegenForOverloadedFunctions(fns: FunctionDeclaration[]): string {
  return [
    codegenForSingleFunction(fns[0]),
    ...fns.map(f => codegenForSingleFunction(f, `'${getSignatureForFn(f)}'`)),
  ].join('\n');
}

function codegenForSingleFunction(fn: FunctionDeclaration, overloadedName?: string): string {
  const methodType = fn.stateMutability === 'view' ? 'CallMethod' : 'SendMethod';
  return `${overloadedName ?? fn.name}: ${methodType}<${codegenInputTypes(
    fn.inputs,
  )}, ${codegenOutputTypes(fn.outputs)}>;`;
}
