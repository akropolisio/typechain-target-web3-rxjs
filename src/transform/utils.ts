import { O } from 'ts-toolbelt';
import {
  AbiOutputParameter,
  AbiParameter,
  EventDeclaration,
  EvmOutputType,
  FunctionDeclaration,
  getSignatureForFn,
  TupleType,
} from 'typechain';

export function unfoldDeclarations<T extends FunctionDeclaration | EventDeclaration>(
  declaration: Record<string, T[]>,
): T[] {
  const toOverloadedDeclaration = (entry: T) => ({
    ...entry,
    name: `"${getSignatureForFn(entry as FunctionDeclaration)}"`,
  });

  return Object.values(declaration).flatMap(entries =>
    entries.length === 1 ? entries[0] : [entries[0], ...entries.map(toOverloadedDeclaration)],
  );
}

export function codegenInputTypes(inputs: O.Optional<AbiParameter, 'name'>[]): string {
  if (inputs.length === 0) {
    return 'void';
  }
  return `{
    ${inputs
      .map((input, index) => `${input.name || `arg${index}`}: ${codegenType(input.type)}`)
      .join(';\n')}
  }`;
}

export function codegenOutputTypes(outputs: O.Optional<AbiOutputParameter, 'name'>[]): string {
  if (outputs.length === 1) {
    return codegenType(outputs[0].type);
  }
  return `{
      ${outputs.map(t => t.name && `${t.name}: ${codegenType(t.type)};\n`).join('')}
      ${outputs.map((t, i) => `${i}: ${codegenType(t.type)}`).join(';\n')}
    }`;
}

export function codegenType(evmType: EvmOutputType): string {
  switch (evmType.type) {
    case 'integer':
    case 'uinteger':
      return 'BN';
    case 'address':
      return 'string';
    case 'bytes':
    case 'dynamic-bytes':
      return 'string';
    case 'array':
      return `(${codegenType(evmType.itemType)})[]`;
    case 'boolean':
      return 'boolean';
    case 'string':
      return 'string';
    case 'tuple':
      return codegenTupleType(evmType);
    case 'void':
      return 'void';
    case 'unknown':
    default:
      throw new Error(`Unhandled type while code generation: ${evmType.originalType}`);
  }
}

export function codegenTupleType(tuple: TupleType) {
  return `{
    ${tuple.components
      .map(component => `${component.name}: ${codegenType(component.type)}`)
      .join(';\n')}
  }`;
}

export function uniq<T>(array: T[]): T[] {
  return array.reduce((acc, curr) => (acc.includes(curr) ? acc : [...acc, curr]), [] as T[]);
}
