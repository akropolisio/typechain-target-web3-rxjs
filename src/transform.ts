/* eslint-disable prettier/prettier */
/* eslint-disable import/no-extraneous-dependencies */
import {
  Contract,
  AbiParameter,
  FunctionDeclaration,
  EventDeclaration,
  AbiOutputParameter,
} from 'typechain';

export function transform(contract: Contract) {
  const name = contract.name.split('.')[0];
  const Name = `${name.slice(0, 1).toUpperCase()}${name.slice(1)}`;

  const events = Object.values(contract.events).map(v => v[0]);
  const methods = Object.values(contract.functions).map(v => v[0]);
  const callMethods = methods.filter(m => m.stateMutability === 'view');
  const sendMethods = methods.filter(m => m.stateMutability !== 'view');

  const template = `
import { makeContractCreator, getOutput, getInput } from './utils/makeContractCreator';

import ${name} from './abi/${contract.name}';

export const create${Name} = makeContractCreator(
  ${name} as any[],
  {
    callMethods: {
      ${callMethods.map(generateFunction).join('\n')}
    },
    sendMethods: {
      ${sendMethods.map(generateFunction).join('\n')}
    },
    events: {
      ${events.map(generateEvent).join('\n')}
    },
  } as const,
);
  `;

  return template;
}

function generateFunction(fn: FunctionDeclaration): string {
  return `
    ${fn.name}: {
      ${
        fn.inputs.length
          ? `inputs: [${fn.inputs.map(x => generateIO('input', x)).join(', ')}],`
          : ''
      }
      ${fn.outputs[0] ? `output: [${fn.outputs.map(x => generateIO('output', x)).join(', ')}]` : ''}
  },`;
}

function generateEvent(event: EventDeclaration) {
  return `
    ${event.name}: {
      inputs: [${event.inputs.map(x => generateIO('input', x)).join(', ')}],
    },
  `;
}

const ioFuncByType = {
  input: 'getInput',
  output: 'getOutput',
};

type ParamByIOType = {
  input: AbiParameter;
  output: AbiOutputParameter;
};

function generateIO<T extends 'input' | 'output'>(type: T, param: ParamByIOType[T]) {
  const params: string[] = [
    type === 'input' ? `'${param.name}'` : '',
    `'${param.type.type === 'array' ? param.type.itemType.type : param.type.type}'`,
    param.type.type === 'array' ? 'true' : '',
  ].filter(Boolean);

  return `${ioFuncByType[type]}(${params.join(', ')})`;
}

// // eslint-disable-next-line consistent-return
// function generateInputType(evmType: EvmType): string {
//   // eslint-disable-next-line default-case
//   switch (evmType.type) {
//     case 'integer':
//     case 'uinteger':
//       return 'number | string';
//     case 'address':
//       return 'string';
//     case 'bytes':
//     case 'dynamic-bytes':
//       return 'string | number[]';
//     case 'array':
//       return `(${generateInputType(evmType.itemType)})[]`;
//     case 'boolean':
//       return 'boolean';
//     case 'string':
//       return 'string';
//     case 'tuple':
//       return generateTupleType(evmType, generateInputType);
//   }
// }

// // eslint-disable-next-line consistent-return
// function generateOutputType(evmType: EvmOutputType): string {
//   // eslint-disable-next-line default-case
//   switch (evmType.type) {
//     case 'integer':
//       return 'BN';
//     case 'uinteger':
//       return 'BN';
//     case 'address':
//       return 'string';
//     case 'void':
//       return 'void';
//     case 'bytes':
//     case 'dynamic-bytes':
//       return 'string';
//     case 'array':
//       return `(${generateOutputType(evmType.itemType)})[]`;
//     case 'boolean':
//       return 'boolean';
//     case 'string':
//       return 'string';
//     case 'tuple':
//       return generateTupleType(evmType, generateOutputType);
//   }
// }

// function generateTupleType(tuple: TupleType, generator: (evmType: EvmType) => string) {
//   return `{${tuple.components
//     .map(component => `${component.name}: ${generator(component.type)}`)
//     .join(', ')}}`;
// }
