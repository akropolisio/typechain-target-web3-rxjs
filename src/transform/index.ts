/* eslint-disable prettier/prettier */
/* eslint-disable import/no-extraneous-dependencies */
import { Contract } from 'typechain';

import { codegenEvents, codegenPastEvents } from './events';
import { codegenMethods } from './methods';

export function transform(contract: Contract) {
  const contractName = `${contract.name.slice(0, 1).toLowerCase()}${contract.name.slice(1)}`;
  const name = contractName.split('.')[0];
  const Name = `${name.slice(0, 1).toUpperCase()}${name.slice(1)}`;

  // const methods = Object.values(contract.functions).map(v => v[0]);

  const template = `
import * as utils from './utils/makeContractCreator';
import { ContractWrapper } from './utils/types';

import ${name} from './abi/${contractName}';

export const create${Name} = utils.makeContractCreator<>(${name} as any[]);

interface ${Name}Contract extends ContractWrapper {
  methods: {
    ${codegenMethods(contract.functions)}
  };
  events: {
    ${codegenEvents(contract.events)}
  };
  getPastEvents: {
    ${codegenPastEvents(contract.events)}
  };
}
  `;

  return template;
}

/* 
interface TestContract extends ContractWrapper {
  methods: {
    testRead: CallMethod<{ s: { a: boolean; b: string[]; c: { x: BN; y: BN }[] }; t: { x: string; y: string }, a: BN }, void>;
    'testRead((boolean,address[],(uint256,uint256)[]),(bytes,string),uint256)': CallMethod<{ s: { a: boolean; b: string[]; c: { x: BN; y: BN }[] }; t: { x: string; y: string }, a: BN }, void>;
    'testRead()': CallMethod<void, void>;
    testSend: SendMethod<{ s: { a: BN; b: BN[]; c { x: BN; y: BN }[] }; t: { x: string; y: string }, a: BN }, boolean>;
  };
  events: {
    TestEvent: EventMethod<EventInput>;
    TestEvent1: EventMethod<Event1Input>;
    TestEvent2: EventMethod<Event2Input>;
    allEvents: EventMethod<EventInput & Event1Input & Event2Input>;
  };
  getPastEvents: PastEventsMethod<{
    TestEvent: EventInput;
    TestEvent1: Event1Input;
    TestEvent2: Event2Input;
  }>;
}
*/
