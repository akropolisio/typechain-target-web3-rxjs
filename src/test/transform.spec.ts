import { parse } from 'typechain';

import { transform } from 'transform';

describe('transform', () => {
  it('should throw error when unknown input type occurs', () => {
    const errorAbi = [
      {
        name: 'testError',
        type: 'function',
        stateMutability: 'view',
        inputs: [
          { name: 'a', type: 'string' },
          { name: 'b', type: 'NotSupported' },
        ],
        outputs: [],
      },
    ];
    const contract = parse(errorAbi as any, 'errorContract');
    const errorTransform = () => transform(contract);
    expect(errorTransform).toThrowErrorMatchingSnapshot();
  });

  it('should throw error when unknown output type occurs', () => {
    const errorAbi = [
      {
        name: 'testError',
        type: 'function',
        stateMutability: 'view',
        inputs: [],
        outputs: [
          {
            name: '',
            type: 'tuple',
            components: [
              {
                name: '',
                type: 'NotSupported',
              },
              {
                name: '',
                type: 'address',
              },
            ],
          },
        ],
      },
    ];
    const contract = parse(errorAbi as any, 'errorContract');
    const errorTransform = () => transform(contract);
    expect(errorTransform).toThrowErrorMatchingSnapshot();
  });
});
