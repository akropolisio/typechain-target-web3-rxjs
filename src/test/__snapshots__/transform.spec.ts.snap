// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform should do anything 1`] = `
Object {
  "contract": Object {
    "constructor": Array [],
    "documentation": undefined,
    "events": Object {
      "Event1": Array [
        Object {
          "inputs": Array [
            Object {
              "isIndexed": true,
              "name": "b",
              "type": Object {
                "bits": 256,
                "originalType": "uint256",
                "type": "uinteger",
              },
            },
            Object {
              "isIndexed": false,
              "name": "c",
              "type": Object {
                "originalType": "bytes32",
                "size": 32,
                "type": "bytes",
              },
            },
          ],
          "isAnonymous": false,
          "name": "Event1",
        },
        Object {
          "inputs": Array [
            Object {
              "isIndexed": true,
              "name": "b",
              "type": Object {
                "bits": 256,
                "originalType": "uint256",
                "type": "uinteger",
              },
            },
            Object {
              "isIndexed": false,
              "name": "c",
              "type": Object {
                "itemType": Object {
                  "components": Array [
                    Object {
                      "name": "x",
                      "type": Object {
                        "bits": 256,
                        "originalType": "uint256",
                        "type": "uinteger",
                      },
                    },
                    Object {
                      "name": "y",
                      "type": Object {
                        "bits": 256,
                        "originalType": "uint256",
                        "type": "uinteger",
                      },
                    },
                  ],
                  "originalType": "tuple",
                  "type": "tuple",
                },
                "originalType": "tuple[]",
                "type": "array",
              },
            },
          ],
          "isAnonymous": false,
          "name": "Event1",
        },
      ],
      "Event2": Array [
        Object {
          "inputs": Array [
            Object {
              "isIndexed": true,
              "name": "b",
              "type": Object {
                "bits": 256,
                "originalType": "uint256",
                "type": "uinteger",
              },
            },
            Object {
              "isIndexed": false,
              "name": "c",
              "type": Object {
                "originalType": "bytes32",
                "size": 32,
                "type": "bytes",
              },
            },
          ],
          "isAnonymous": false,
          "name": "Event2",
        },
      ],
    },
    "fallback": undefined,
    "functions": Object {
      "testRead": Array [
        Object {
          "documentation": undefined,
          "inputs": Array [
            Object {
              "name": "s",
              "type": Object {
                "components": Array [
                  Object {
                    "name": "a",
                    "type": Object {
                      "originalType": "bool",
                      "type": "boolean",
                    },
                  },
                  Object {
                    "name": "b",
                    "type": Object {
                      "itemType": Object {
                        "originalType": "address",
                        "type": "address",
                      },
                      "originalType": "address[]",
                      "type": "array",
                    },
                  },
                  Object {
                    "name": "c",
                    "type": Object {
                      "itemType": Object {
                        "components": Array [
                          Object {
                            "name": "x",
                            "type": Object {
                              "bits": 256,
                              "originalType": "uint256",
                              "type": "uinteger",
                            },
                          },
                          Object {
                            "name": "y",
                            "type": Object {
                              "bits": 256,
                              "originalType": "uint256",
                              "type": "uinteger",
                            },
                          },
                        ],
                        "originalType": "tuple",
                        "type": "tuple",
                      },
                      "originalType": "tuple[]",
                      "type": "array",
                    },
                  },
                ],
                "originalType": "tuple",
                "type": "tuple",
              },
            },
            Object {
              "name": "t",
              "type": Object {
                "components": Array [
                  Object {
                    "name": "x",
                    "type": Object {
                      "originalType": "bytes",
                      "type": "dynamic-bytes",
                    },
                  },
                  Object {
                    "name": "y",
                    "type": Object {
                      "originalType": "string",
                      "type": "string",
                    },
                  },
                ],
                "originalType": "tuple",
                "type": "tuple",
              },
            },
            Object {
              "name": "a",
              "type": Object {
                "bits": 256,
                "originalType": "uint256",
                "type": "uinteger",
              },
            },
          ],
          "name": "testRead",
          "outputs": Array [
            Object {
              "name": "",
              "type": Object {
                "originalType": "address",
                "type": "address",
              },
            },
            Object {
              "name": "Test",
              "type": Object {
                "originalType": "address",
                "type": "address",
              },
            },
            Object {
              "name": "",
              "type": Object {
                "originalType": "bool",
                "type": "boolean",
              },
            },
          ],
          "stateMutability": "view",
        },
        Object {
          "documentation": undefined,
          "inputs": Array [],
          "name": "testRead",
          "outputs": Array [
            Object {
              "name": "",
              "type": Object {
                "bits": 256,
                "originalType": "uint256",
                "type": "uinteger",
              },
            },
          ],
          "stateMutability": "view",
        },
      ],
      "testSend": Array [
        Object {
          "documentation": undefined,
          "inputs": Array [
            Object {
              "name": "s",
              "type": Object {
                "components": Array [
                  Object {
                    "name": "a",
                    "type": Object {
                      "bits": 256,
                      "originalType": "uint256",
                      "type": "uinteger",
                    },
                  },
                  Object {
                    "name": "b",
                    "type": Object {
                      "itemType": Object {
                        "bits": 256,
                        "originalType": "uint256",
                        "type": "uinteger",
                      },
                      "originalType": "uint256[]",
                      "type": "array",
                    },
                  },
                  Object {
                    "name": "c",
                    "type": Object {
                      "itemType": Object {
                        "components": Array [
                          Object {
                            "name": "x",
                            "type": Object {
                              "bits": 256,
                              "originalType": "uint256",
                              "type": "uinteger",
                            },
                          },
                          Object {
                            "name": "y",
                            "type": Object {
                              "bits": 256,
                              "originalType": "uint256",
                              "type": "uinteger",
                            },
                          },
                        ],
                        "originalType": "tuple",
                        "type": "tuple",
                      },
                      "originalType": "tuple[]",
                      "type": "array",
                    },
                  },
                ],
                "originalType": "tuple",
                "type": "tuple",
              },
            },
            Object {
              "name": "t",
              "type": Object {
                "components": Array [
                  Object {
                    "name": "x",
                    "type": Object {
                      "bits": 256,
                      "originalType": "uint256",
                      "type": "uinteger",
                    },
                  },
                  Object {
                    "name": "y",
                    "type": Object {
                      "bits": 256,
                      "originalType": "uint256",
                      "type": "uinteger",
                    },
                  },
                ],
                "originalType": "tuple",
                "type": "tuple",
              },
            },
            Object {
              "name": "a",
              "type": Object {
                "bits": 256,
                "originalType": "uint256",
                "type": "uinteger",
              },
            },
          ],
          "name": "testSend",
          "outputs": Array [
            Object {
              "name": "",
              "type": Object {
                "originalType": "bool",
                "type": "boolean",
              },
            },
          ],
          "stateMutability": "nonpayable",
        },
      ],
    },
    "name": "Test",
    "rawName": "test",
    "structs": Object {
      "undefined": Array [
        Object {
          "itemType": Object {
            "originalType": "address",
            "type": "address",
          },
          "originalType": "address[]",
          "type": "array",
        },
      ],
    },
  },
  "result": "
import * as utils from './utils/makeContractCreator';
import { ContractWrapper } from './utils/types';

import test from './abi/test';

export const createTest = utils.makeContractCreator<>(test as any[]);

interface TestContract extends ContractWrapper {
  methods: {
    testRead: CallMethod<{
    s: {
    a: boolean;
b: (string)[];
c: ({
    x: BN;
y: BN
  })[]
  };
t: {
    x: string;
y: string
  };
a: BN
  }, {
      Test: string;

      0: string;
1: string;
2: boolean
    }>;
'testRead((bool,address[],(uint256,uint256)[]),(bytes,string),uint256)': CallMethod<{
    s: {
    a: boolean;
b: (string)[];
c: ({
    x: BN;
y: BN
  })[]
  };
t: {
    x: string;
y: string
  };
a: BN
  }, {
      Test: string;

      0: string;
1: string;
2: boolean
    }>;
'testRead()': CallMethod<void, BN>;
testSend: SendMethod<{
    s: {
    a: BN;
b: (BN)[];
c: ({
    x: BN;
y: BN
  })[]
  };
t: {
    x: BN;
y: BN
  };
a: BN
  }, boolean>;
  };
  events: {
    Event1: EventMethod<
    {
    b: BN;
c: string
  },
    {
    b: BN
  }
  >
'Event1(uint256,bytes32)': EventMethod<
    {
    b: BN;
c: string
  },
    {
    b: BN
  }
  >
'Event1(uint256,(uint256,uint256)[])': EventMethod<
    {
    b: BN;
c: ({
    x: BN;
y: BN
  })[]
  },
    {
    b: BN
  }
  >
Event2: EventMethod<
    {
    b: BN;
c: string
  },
    {
    b: BN
  }
  >
  };
  getPastEvents: {
    Event1;
Event1;
Event2
  };
}
  ",
}
`;
